R/utils.R

```diff
@@ -38,6 +38,105 @@ h_get_vars <- function(treatment) {
+#' - `n_levels`: Number of treatment levels.
+#' - `levels`: Names of the treatment levels.
+#'
+#' @keywords internal
+h_prep_survival_vars <- function(formula, data, treatment) {
+  trt_vars <- h_get_vars(treatment)
+  assert_data_frame(data)
+  assert_formula(formula)
+  assert_true(identical(length(formula), 3L))
+  assert_subset(c(trt_vars$treatment, trt_vars$strata), colnames(data))
+  assert_subset(trt_vars$treatment, all.vars(formula[[3]]))
+  assert(
+    test_character(data[[trt_vars$treatment]]),
+    test_factor(data[[trt_vars$treatment]])
+  )
+
+  trt_lvls <- levels(data[[trt_vars$treatment]])
+  n_lvls <- length(trt_lvls)
+  covariates <- setdiff(all.vars(formula[[3]]), c(trt_vars$treatment, trt_vars$strata))
+
+  # Extract survival time and censoring indicator from the left hand side of the formula.
+  lhs <- formula[[2]]
+  if (inherits(lhs, "call") && (lhs[[1]] == as.name("Surv") || lhs[[1]] == as.name("survival::Surv"))) {
+    surv_vars <- as.character(lhs)[-1]
+    assert_subset(surv_vars, colnames(data))
+    time_var <- surv_vars[1]
+    status_var <- surv_vars[2]
+  } else {
+    stop("Left hand side of formula must be a Surv() object.")
+  }
+
+  # Extract model without left hand side and without treatment and strata variables.
+  model <- as.formula(as.call(as.list(formula)[-2L]))
+  update_string <- if (length(trt_vars$strata)) {
+    paste0("~ . - ", trt_vars$treatment, "*", trt_vars$strata)
+  } else {
+    paste("~ . -", trt_vars$treatment)
+  }
+  model <- update(model, as.formula(update_string))
+
+  list(
+    time = time_var,
+    status = status_var,
+    treatment = trt_vars$treatment,
+    strata = trt_vars$strata,
+    schema = trt_vars$schema,
+    covariates = covariates,
+    model = model,
+    n_levels = n_lvls,
+    levels = trt_lvls
+  )
+}
+
+#' Count Number of Events per Unique Event Time
+#'
+#' This function counts the number of events at each unique event time point in a survival dataset.
+#'
+#' @details If there are no events in the dataset, it returns an empty `data.frame`.
+#'
+#' @param df (`data.frame`) containing the survival data.
+#' @param time (`string`) name of the time variable.
+#' @param status (`string`) name of the status variable, where 1 indicates an event and 0 indicates censoring.
+#' @return A `data.frame` with two columns: `time` and `n_events`, where `n_events` is the
+#'   number of events at each time point.
+#' @keywords internal
+h_n_events_per_time <- function(df, time, status) {
+  assert_data_frame(df)
+  assert_string(time)
+  assert_string(status)
+  assert_numeric(df[[time]], any.missing = FALSE)
+  assert_numeric(df[[status]], any.missing = FALSE)
+  assert_true(all(df[[status]] %in% c(0, 1)))
+
+  has_event <- df[[status]] == 1
+  df_events <- df[has_event, , drop = FALSE]
+  if (nrow(df_events) == 0) {
+    return(data.frame(time = numeric(0), n_events = integer(0)))
+  }
+  times_count <- table(df_events[[time]])
```

Comment:
how about using `aggregate` function? in table it convert the time from numeric to strings so feel a bit concerned about this conversion